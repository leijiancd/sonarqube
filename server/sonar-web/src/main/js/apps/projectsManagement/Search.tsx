/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { sortBy } from 'lodash';
import BulkApplyTemplateModal from './BulkApplyTemplateModal';
import DeleteModal from './DeleteModal';
import { Type, QUALIFIERS_ORDER } from './utils';
import { Project } from './utils';
import { Organization } from '../../app/types';
import RadioToggle from '../../components/controls/RadioToggle';
import Checkbox from '../../components/controls/Checkbox';
import { translate } from '../../helpers/l10n';

export interface Props {
  onAllDeselected: () => void;
  onAllSelected: () => void;
  onDeleteProjects: () => void;
  onQualifierChanged: (qualifier: string) => void;
  onSearch: (query: string) => void;
  onTypeChanged: (type: Type) => void;
  organization: Organization;
  projects: Project[];
  qualifiers: string;
  query: string;
  ready: boolean;
  selection: any[];
  topLevelQualifiers: string[];
  total: number;
  type: Type;
}

interface State {
  bulkApplyTemplateModal: boolean;
  deleteModal: boolean;
}

export default class Search extends React.PureComponent<Props, State> {
  input: HTMLInputElement;
  mounted: boolean;
  state: State = { bulkApplyTemplateModal: false, deleteModal: false };

  onSubmit = (event: React.SyntheticEvent<HTMLFormElement>) => {
    event.preventDefault();
    this.search();
  };

  search = (event?: React.SyntheticEvent<HTMLInputElement>) => {
    const q = event ? event.currentTarget.value : this.input.value;
    this.props.onSearch(q);
  };

  getTypeOptions = () => [
    { value: Type.All, label: 'All' },
    { value: Type.Provisioned, label: 'Provisioned' }
  ];

  getQualifierOptions = () => {
    const options = this.props.topLevelQualifiers.map(q => {
      return { value: q, label: translate('qualifiers', q) };
    });
    return sortBy(options, option => QUALIFIERS_ORDER.indexOf(option.value));
  };

  onCheck = (checked: boolean) => {
    if (checked) {
      this.props.onAllSelected();
    } else {
      this.props.onAllDeselected();
    }
  };

  handleDeleteClick = (event: React.SyntheticEvent<HTMLButtonElement>) => {
    event.preventDefault();
    event.currentTarget.blur();
    this.setState({ deleteModal: true });
  };

  closeDeleteModal = () => {
    this.setState({ deleteModal: false });
  };

  handleDeleteConfirm = () => {
    this.closeDeleteModal();
    this.props.onDeleteProjects();
  };

  handleBulkApplyTemplateClick = (event: React.SyntheticEvent<HTMLButtonElement>) => {
    event.preventDefault();
    event.currentTarget.blur();
    this.setState({ bulkApplyTemplateModal: true });
  };

  closeBulkApplyTemplateModal = () => {
    this.setState({ bulkApplyTemplateModal: false });
  };

  renderCheckbox = () => {
    const isAllChecked =
      this.props.projects.length > 0 && this.props.selection.length === this.props.projects.length;
    const thirdState =
      this.props.projects.length > 0 &&
      this.props.selection.length > 0 &&
      this.props.selection.length < this.props.projects.length;
    const checked = isAllChecked || thirdState;
    return <Checkbox checked={checked} thirdState={thirdState} onCheck={this.onCheck} />;
  };

  renderQualifierFilter = () => {
    const options = this.getQualifierOptions();
    if (options.length < 2) {
      return null;
    }
    return (
      <td className="thin nowrap text-middle">
        <RadioToggle
          options={this.getQualifierOptions()}
          value={this.props.qualifiers}
          name="projects-qualifier"
          onCheck={this.props.onQualifierChanged}
        />
      </td>
    );
  };

  render() {
    const isSomethingSelected = this.props.projects.length > 0 && this.props.selection.length > 0;
    return (
      <div className="panel panel-vertical bordered-bottom spacer-bottom">
        <table className="data">
          <tbody>
            <tr>
              <td className="thin text-middle">
                {this.props.ready ? this.renderCheckbox() : <i className="spinner" />}
              </td>
              {this.renderQualifierFilter()}
              <td className="thin nowrap text-middle">
                <RadioToggle
                  options={this.getTypeOptions()}
                  value={this.props.type}
                  name="projects-type"
                  onCheck={this.props.onTypeChanged}
                />
              </td>
              <td className="text-middle">
                <form onSubmit={this.onSubmit} className="search-box">
                  <button className="search-box-submit button-clean">
                    <i className="icon-search" />
                  </button>
                  <input
                    onChange={this.search}
                    value={this.props.query}
                    ref={node => (this.input = node!)}
                    className="search-box-input input-medium"
                    type="search"
                    placeholder="Search"
                  />
                </form>
              </td>
              <td className="thin nowrap text-middle">
                <button
                  className="spacer-right js-bulk-apply-permission-template"
                  onClick={this.handleBulkApplyTemplateClick}>
                  {translate('permission_templates.bulk_apply_permission_template')}
                </button>
                <button
                  onClick={this.handleDeleteClick}
                  className="js-delete button-red"
                  disabled={!isSomethingSelected}>
                  {translate('delete')}
                </button>
              </td>
            </tr>
          </tbody>
        </table>

        {this.state.bulkApplyTemplateModal &&
          <BulkApplyTemplateModal
            onClose={this.closeBulkApplyTemplateModal}
            organization={this.props.organization.key}
            qualifier={this.props.qualifiers}
            query={this.props.query}
            selection={this.props.selection}
            total={this.props.total}
            type={this.props.type}
          />}

        {this.state.deleteModal &&
          <DeleteModal
            onClose={this.closeDeleteModal}
            onConfirm={this.handleDeleteConfirm}
            organization={this.props.organization.key}
            qualifier={this.props.qualifiers}
            selection={this.props.selection}
          />}
      </div>
    );
  }
}
