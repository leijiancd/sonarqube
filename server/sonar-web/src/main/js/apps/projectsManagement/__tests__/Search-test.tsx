/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import Search, { Props } from '../Search';
import { Type } from '../utils';
import { change, click } from '../../../helpers/testUtils';

const organization = { key: 'org', name: 'org', projectVisibility: 'public' };

it('renders', () => {
  expect(shallowRender()).toMatchSnapshot();
});

it('render qualifiers filter', () => {
  expect(shallowRender({ topLevelQualifiers: ['TRK', 'VW', 'APP'] })).toMatchSnapshot();
});

it('updates qualifier', () => {
  const onQualifierChanged = jest.fn();
  const wrapper = shallowRender({ onQualifierChanged, topLevelQualifiers: ['TRK', 'VW', 'APP'] });
  wrapper.find('RadioToggle[name="projects-qualifier"]').prop<Function>('onCheck')('VW');
  expect(onQualifierChanged).toBeCalledWith('VW');
});

it('updates type', () => {
  const onTypeChanged = jest.fn();
  const wrapper = shallowRender({ onTypeChanged });
  wrapper.find('RadioToggle[name="projects-type"]').prop<Function>('onCheck')(Type.Provisioned);
  expect(onTypeChanged).toBeCalledWith(Type.Provisioned);
});

it('searches', () => {
  const onSearch = jest.fn();
  const wrapper = shallowRender({ onSearch });
  change(wrapper.find('input[type="search"]'), 'foo');
  expect(onSearch).toBeCalledWith('foo');
});

it('checks all or none projects', () => {
  const onAllDeselected = jest.fn();
  const onAllSelected = jest.fn();
  const wrapper = shallowRender({ onAllDeselected, onAllSelected });

  wrapper.find('Checkbox').prop<Function>('onCheck')(true);
  expect(onAllSelected).toBeCalled();

  wrapper.find('Checkbox').prop<Function>('onCheck')(false);
  expect(onAllDeselected).toBeCalled();
});

it('deletes projects', () => {
  const onDeleteProjects = jest.fn();
  const wrapper = shallowRender({ onDeleteProjects, selection: ['foo', 'bar'] });
  click(wrapper.find('.js-delete'));
  expect(wrapper.find('DeleteModal')).toMatchSnapshot();
  wrapper.find('DeleteModal').prop<Function>('onConfirm')();
  expect(onDeleteProjects).toBeCalled();
});

it('bulk applies permission template', () => {
  const wrapper = shallowRender({});
  click(wrapper.find('.js-bulk-apply-permission-template'));
  expect(wrapper.find('BulkApplyTemplateModal')).toMatchSnapshot();
  wrapper.find('BulkApplyTemplateModal').prop<Function>('onClose')();
  expect(wrapper.find('BulkApplyTemplateModal').exists()).toBeFalsy();
});

function shallowRender(props?: { [P in keyof Props]?: Props[P] }) {
  return shallow(
    <Search
      onAllDeselected={jest.fn()}
      onAllSelected={jest.fn()}
      onDeleteProjects={jest.fn()}
      onQualifierChanged={jest.fn()}
      onSearch={jest.fn()}
      onTypeChanged={jest.fn()}
      organization={organization}
      projects={[]}
      qualifiers="TRK"
      query=""
      ready={true}
      selection={[]}
      topLevelQualifiers={['TRK']}
      total={0}
      type={Type.All}
      {...props}
    />
  );
}
