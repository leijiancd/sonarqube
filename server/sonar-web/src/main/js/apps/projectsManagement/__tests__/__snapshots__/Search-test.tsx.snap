// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bulk applies permission template 1`] = `
<BulkApplyTemplateModal
  onClose={[Function]}
  organization="org"
  qualifier="TRK"
  query=""
  selection={Array []}
  total={0}
  type="ALL"
/>
`;

exports[`deletes projects 1`] = `
<DeleteModal
  onClose={[Function]}
  onConfirm={[Function]}
  organization="org"
  qualifier="TRK"
  selection={
    Array [
      "foo",
      "bar",
    ]
  }
/>
`;

exports[`render qualifiers filter 1`] = `
<div
  className="panel panel-vertical bordered-bottom spacer-bottom"
>
  <table
    className="data"
  >
    <tbody>
      <tr>
        <td
          className="thin text-middle"
        >
          <Checkbox
            checked={false}
            onCheck={[Function]}
            thirdState={false}
          />
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <RadioToggle
            disabled={false}
            name="projects-qualifier"
            onCheck={[Function]}
            options={
              Array [
                Object {
                  "label": "qualifiers.TRK",
                  "value": "TRK",
                },
                Object {
                  "label": "qualifiers.VW",
                  "value": "VW",
                },
                Object {
                  "label": "qualifiers.APP",
                  "value": "APP",
                },
              ]
            }
            value="TRK"
          />
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <RadioToggle
            disabled={false}
            name="projects-type"
            onCheck={[Function]}
            options={
              Array [
                Object {
                  "label": "All",
                  "value": "ALL",
                },
                Object {
                  "label": "Provisioned",
                  "value": "PROVISIONED",
                },
              ]
            }
            value="ALL"
          />
        </td>
        <td
          className="text-middle"
        >
          <form
            className="search-box"
            onSubmit={[Function]}
          >
            <button
              className="search-box-submit button-clean"
            >
              <i
                className="icon-search"
              />
            </button>
            <input
              className="search-box-input input-medium"
              onChange={[Function]}
              placeholder="Search"
              type="search"
              value=""
            />
          </form>
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <button
            className="spacer-right js-bulk-apply-permission-template"
            onClick={[Function]}
          >
            permission_templates.bulk_apply_permission_template
          </button>
          <button
            className="js-delete button-red"
            disabled={true}
            onClick={[Function]}
          >
            delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`renders 1`] = `
<div
  className="panel panel-vertical bordered-bottom spacer-bottom"
>
  <table
    className="data"
  >
    <tbody>
      <tr>
        <td
          className="thin text-middle"
        >
          <Checkbox
            checked={false}
            onCheck={[Function]}
            thirdState={false}
          />
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <RadioToggle
            disabled={false}
            name="projects-type"
            onCheck={[Function]}
            options={
              Array [
                Object {
                  "label": "All",
                  "value": "ALL",
                },
                Object {
                  "label": "Provisioned",
                  "value": "PROVISIONED",
                },
              ]
            }
            value="ALL"
          />
        </td>
        <td
          className="text-middle"
        >
          <form
            className="search-box"
            onSubmit={[Function]}
          >
            <button
              className="search-box-submit button-clean"
            >
              <i
                className="icon-search"
              />
            </button>
            <input
              className="search-box-input input-medium"
              onChange={[Function]}
              placeholder="Search"
              type="search"
              value=""
            />
          </form>
        </td>
        <td
          className="thin nowrap text-middle"
        >
          <button
            className="spacer-right js-bulk-apply-permission-template"
            onClick={[Function]}
          >
            permission_templates.bulk_apply_permission_template
          </button>
          <button
            className="js-delete button-red"
            disabled={true}
            onClick={[Function]}
          >
            delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
